shelley/chain-and-ledger/dependencies/non-integer/src/NonIntegral.hs

```diff
@@ -5,63 +5,64 @@ module NonIntegral
   , splitLn
   , scaleExp
-  , CompareResult(..)
   , taylorExpCmp
+  , CompareResult(..)
   ) where
 
 data CompareResult a = BELOW a Int
                      | ABOVE a Int
                      | UNKNOWN
   deriving (Show, Eq)
 
-scaleExp :: (RealFrac b) => b -> (Integer, b)
-scaleExp x = (ceiling x, x / fromIntegral (ceiling x :: Integer))
+epsilon :: (Fractional a) => a
+epsilon = 1 / 10^(24::Int)
 
 -- | Exponentiation
 (***) :: (RealFrac a, Enum a, Show a) => a -> a -> a
 a *** b
-  | a == 0 = if b == 0 then 1 else 0
+  | b == 0 = 1
+  | a == 0 = 0
   | a == 1 = 1
-  | otherwise = exp' l
-    where l = b * ln' a
+  | otherwise = exp' (b * ln' a)
 
 ipow' :: Num a => a -> Integer -> a
 ipow' x n
-    | n == 0     = 1
-    | mod n 2 == 0 = let res = ipow' x (div n 2) in res * res
-    | otherwise  = x * ipow' x (n - 1)
+  | n == 0 = 1
+  | m == 0 = let y = ipow' x d in y * y
+  | otherwise = x * ipow' x (n - 1)
+  where (d,m) = divMod n 2
 
 ipow :: Fractional a => a -> Integer -> a
 ipow x n
-    | n < 0 = 1 / (ipow x (-n))
-    | otherwise = ipow' x n
+  | n < 0 = 1 / (ipow' x (-n))
+  | otherwise = ipow' x n
 
 logAs :: (Num a) => a -> [a]
-logAs a = a' : a' : logAs (a + 1)
-  where
-    a' = a * a
+logAs a = a2 : a2 : logAs (a + 1)
+    where a2 = a * a
 
 -- | Approximate ln(1+x) for x \in [0, \infty)
-fln :: (Fractional a, Enum a, Ord a, Show a) => Int -> a -> a
-fln maxN x = if x < 0
-             then error ("x = " ++ show x ++ "is not inside domain [0, ..) ")
-             else cf maxN 0 eps Nothing 1 0 0 1 (x : [a * x | a <- logAs 1]) [1,2 ..]
-
-eps :: (Fractional a) => a
-eps = 1 / 10^(24::Int)
+-- a_{1} = x, a_{2k} = a_{2k+1} = x·k^2, k >= 1
+-- b_{n} = n, n >= 0
+lncf :: (Fractional a, Enum a, Ord a, Show a) => Int -> a -> a
+lncf maxN x
+  | x < 0 = error ("x = " ++ show x ++ " is not inside domain [0,..)")
+  | otherwise = cf maxN 0 epsilon Nothing 1 0 0 1 xs [1,2..]
+  where xs = map (x*) (1 : logAs 1)
 
 -- | Compute continued fraction using max steps or bounded list of a/b factors.
 -- The 'maxN' parameter gives the maximum recursion depth, 'n' gives the current
 -- rursion depth, 'lastVal' is the optional last value ('Nothing' for the first
--- iteration). 'aNm2', 'bNm2' are a_{m-2} and b_{m-2}, and 'aNm1' / 'bNm1' are
--- a_{m-1} / b_{m-1} respectively, 'an' / 'bn' are lists of succecsive a_n / b_n
--- values for the recurrence relation:
+-- iteration). 'aNm2' / 'bNm2' are A_{n-2} / B_{n-2}, 'aNm1' / 'bNm1' are
+-- A_{n-1} / B_{n-1}, and 'aN' / 'bN' are A_{n} / B_{n} respectively, 'an' / 'bn' 
+-- are lists of succecsive a_{n} / b_{n} values for the recurrence relation:
 --
--- A_{-1}  = 1                              B_{-1} = 0
--- A_0     = b_0                            B_0    = 1
--- A_{n+1} = b_{n+1}*A_n + a_{n+1}*A_{n-1}  B_{n+1} = b_{n+1}*B_n + a_{n+1}*B_{n-1}
+-- A_{-1} = 1,    A_{0} = b_{0}
+-- B_{-1} = 0,    B_{0} = 1
+-- A_{n}  = b_{n}*A_{n-1} + a_{n}*A_{n-2}
+-- B_{n}  = b_{n}*B_{n-1} + a_{n}*B_{n-2}
```

Comment:
please revert this hunk

Reply:
Do you find the other changes in this file worth the trouble? Maybe I should send only the changes int the testing code and leave this one untouched.

Reply:
I have to have a closer look (sorry, a bit short of time for now) but overall I think a better structuring of the code is good.
Undoing a particular part of a commit is easy or hard, depends mainly on the UI you use for git. What you can always do is to "undo" the changes manually, commit the change and then squash with the original commit.